lua/nvim-tree/actions/expand-all.lua

```diff
@@ -20,34 +21,38 @@ local function expand(node)
   end
 end
 
+local function should_expand(expansion_count, node)
+  local should_halt = expansion_count >= M.MAX_FOLDER_DISCOVERY
+  local should_exclude = M.EXCLUDE[node.name]
+  return not should_halt and node.nodes and not node.open and not should_exclude
+end
+
 local function gen_iterator()
   local expansion_count = 0
 
-  local function iterate(parent)
-    if expansion_count >= M.MAX_FOLDER_DISCOVERY then
-      return true
-    end
-
+  return function(parent)
     if parent.parent and parent.nodes and not parent.open then
       expansion_count = expansion_count + 1
       expand(parent)
     end
 
-    for _, node in pairs(parent.nodes) do
-      if node.nodes and not node.open and not M.EXCLUDE[node.name] then
-        expansion_count = expansion_count + 1
-        expand(node)
-      end
-
-      if node.open then
-        if iterate(node) then
-          return true
+    Iterator.new(parent.nodes)
```

Comment:
I love the fluid nature. This is classic builder pattern. Perhaps we could make the naming more concise e.g.

```lua
NodeIterator.builder(parent.nodes)
  :hidden()
  :matcher(...)
  :applier(...)
  :recursor(...)
  :build()()
```

The `:build()()` is a bit redundant and implies that we have an intermediate `NodeIteratorBuilder` which doesn't add much value, so `:consume()` or `iterate()` is appropriate.